### PR TITLE
spmd_types 0.2.2 update

### PR DESCRIPTION
**Pull request by pianpwk**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.12.0) (oldest at bottom):
* __->__ #3936
* #3905

### CODE REVIEW

File: torchtitan/components/loss.py [hunk 1]
```diff
@@ -103,9 +104,14 @@ def typecheck_forward(
             global_vocab_size,
             reduction,
         )
-        output_type = dict(spmd.get_local_type(logits))
-        output_type[tp_group] = spmd.I
-        spmd.assert_type(result, output_type)
+        out_type = spmd.get_local_type(logits)
+        if reduction == "sum":
```
Comment (tianyu-l) [inline torchtitan/components/loss.py:108]: What about `mean`? In fact, can we just ban `mean`?

File: torchtitan/distributed/parallel_dims.py [hunk 2]
```diff
@@ -52,67 +58,39 @@ class MeshAxisName(StrEnum):
     EFSDP = "efsdp"
 
 
-@dataclass(frozen=True, slots=True)
-class SpmdLayout:
-    """Temporary SPMD layout annotations keyed by logical mesh axis name.
+SpmdLayout = spmd.SpmdType
```
Comment (tianyu-l) [inline torchtitan/distributed/parallel_dims.py:61]: why aliasing?
- if `SpmdType` is the official name we should use it
- if it's not good name, why making it official

File: torchtitan/distributed/parallel_dims.py [hunk 3]
```diff
@@ -52,67 +58,39 @@ class MeshAxisName(StrEnum):
...
     """
+    Return per-axis types with PartitionSpec sharding represented as S(i).
+    e.g. {DP: R, CP: V} + PartitionSpec(None, CP) -> {DP: R, CP: S(1)}
 
-    axis_types: dict[MeshAxisName, spmd.PerMeshAxisSpmdType]
-    partition_spec: spmd.PartitionSpec | tuple[Any, ...] | None = None
+    This is not meant as a minimal description of the SPMD layout; shard order
+    cannot be expressed. Specifically, shard order information will be lost in
+    this representation. This is purely a helper for calling spmd.redistribute,
+    which takes per-axis types (e.g. redistribute(S(1) -> R)).
```
Comment (tianyu-l) [inline torchtitan/distributed/parallel_dims.py:74]: Is this titan specific?

If so, should we put it under `spmd_types.py` (after dtensor deprecation)?
